core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.storage.BlockManagerId
+
+/**
+ * :: Experimental ::
+ * Pluggable interface to get merger locations for Push based shuffle.
+ */
+@Experimental
+@Since("3.1.0")
+trait ShuffleMergerDiscovery {
+
+ /**
+  * Get the list of shuffle services locations in order to push the
+  * shuffle blocks for merge
+  */
+  def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]) : Seq[BlockManagerId]
```

Comment:
So, I'm seeing multiple getMergerLocations API here.
One is from `SchedulerBackend`:
> def getMergerLocations(numPartitions: Int, resourceProfileId: Int): Seq[BlockManagerId]

The 2nd is from `BlockManagerMaster `:
> def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]): Seq[BlockManagerId]

The 3rd is from `ShuffleMergerDiscovery`:
> def getMergerLocations(numMergersNeeded: Int, hostsToFilter: Set[String]) : Seq[BlockManagerId]

DAGScheduler first invokes `getMergerLocations` API from `SchedulerBackend`, which then invokes the `getMergerLocations` API from `BlockManagerMaster`, which then invokes the `getMergerLocations` API from `ShuffleMergerDiscovery`.
This is quite confusing, since when we use the `getMergerLocations` pluggable API, the invocation has already passed through 2 layers.
The pluggable API should be the first level of invocation, which encapsulates the entire invocation path.
`hostsToFilter` shouldn't be a parameter of the API.
It should rather be the internal state managed by the actual implementation.
In the case of the default implementation, the scheduler blacklist provides `hostsToFilter`.
If we implement customized merger coordinator, it would maintain its own state for what hosts to filter.

Reply:
Its added to `SchedulerBackend` initially because different cluster managers can have different ways of getting merger locations. If we don't pass `hostsToFilter` to the API then we might end up with lesser than the required merger locations then we have to make another call which is why I pass this info to the API.

Reply:
After having offline discussions with @Victsm decided to leave it like this and get more feedback from OS before making more changes to it.